streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -234,14 +235,19 @@ void initialize() {
         }
 
         void pollAndUpdate() {
-            final ConsumerRecords<byte[], byte[]> received = globalConsumer.poll(pollMs);
-            for (ConsumerRecord<byte[], byte[]> record : received) {
-                stateMaintainer.update(record);
-            }
-            final long now = time.milliseconds();
-            if (flushInterval >= 0 && now >= lastFlush + flushInterval) {
-                stateMaintainer.flushState();
-                lastFlush = now;
+            try {
+                final ConsumerRecords<byte[], byte[]> received = globalConsumer.poll(pollMs);
+                for (ConsumerRecord<byte[], byte[]> record : received) {
+                    stateMaintainer.update(record);
+                }
+                final long now = time.milliseconds();
+                if (flushInterval >= 0 && now >= lastFlush + flushInterval) {
+                    stateMaintainer.flushState();
+                    lastFlush = now;
+                }
+            } catch (final InvalidOffsetException recoverableException) {
```

Comment:
Hmmm, this i'm not 100% sure of. I believe it will likely result in failures on other threads, i.e., when they try to join to a globalTable that has been deleted or is no longer open.

Reply:
Good point. Will try to come up with a solution (and hopefully a test)

Reply:
I discussed with @guozhangwang about this. It's a little tricky and here we expect an exception only in very rare corner cases (maybe even never in practice).

Thus, one option would be to just die for this case and users need to restart their KafkaStreams instance.

Another alternative would be to add a `clear()` API to stores. Using the `clear()` API, we could just delete the current content, seekToEnd() and continue populated the store (this seems to align with our guarantees about a global store).

Third (in case we don't want to do a KIP as needed for second option), we follow this PR with a slight adjustment -- instead of triggering a restore, we only do "close, wipe out, reopen". This will keep the window when the store is not available very short and would allow us to put some retry logic for stream-globalTKable join in place (in `KTableSourceValueGetter` we just catch `InvalidStateStoreException` backoff and retry until the store is open again). This disadvantage here is, that we need to educate PAPI uses about this behavior and that the need to implement the same pattern.

It seems for now, we can implement (1) and do a follow up KIP to get to (2). Option (3) seems a little bit hacky to me, and the PAPI issues bothers me, too.

WDYT? \cc @bbejeck @guozhangwang

Reply:
One more thought: atm, we use the `storeName` as directory name within the task directory if the store is persistent in our implementation. However, we don't force users implementing a  custom store do follow this pattern. We could enforce this, by passing in the store directory (instead of the task directory) but this might be an incompatible change. Note, with approach (1), we are actually relying on an implementation like this. Thus, we should at least document this requirement.

Please share your thoughts.

Reply:
I guess option 1 seems like a reasonable thing to do right now. With option 2 maybe a future enhancement.

For the directory stuff, yes we should at least document it. Passing in a `storeDirectory` would probably require another public API change? i.e, where would we get it from? I guess we could just use the `StoreBuilder`/`StoreSupplier` `name()` for the directory as we know it will be unique

Reply:
I'd also say go with option 1, especially since it's a corner case.  We can see from testing and user experience if we need to add option 2.

Reply:
I was just thinking about option (2) in combination the store directory issue (ie, use the store name as directory). Even if we force people to use correct store directory, this will not be enough. We need option (2) not just for global store but also for shared stores. The `StateStore` interface is generic and one could also implement a remote store -- for this case, deleting the local store directory is not sufficient and the store need to have a `clear()` API.

Does this make sense? If yes, I'll open a JIRA for this. Not sure how urgent this is to fix, as I am not sure how many people would implement a remote store.

Reply:
Opened https://github.com/apache/kafka/pull/4228 for JavaDoc improvements

Reply:
As discussed offline, I'm in favor of option 1). Since the case of `InvalidOffset` during normal processing would only happen if the global thread is stalled somehow for long enough time that the topic partitions has rolled a new segment such that the old ones can be compacted. Most of time global thread would be at the log tail (i.e. active segment) of the topic partitions so it should be really, really rare.

So to summarize, for the three phases:

1) before restoration, the checkpointed offset is already < logStartOffset,
2) during restoration, the `InvalidOffset` gets returned in `poll()` due to log compaction.
3) during normal processing (for standy tasks and global update tasks), the `InvalidOffset` gets returned in `poll()` due to log compaction.

For local store, we would wipe-and-restore-from-scratch for case 1/2/3; for global store we would wipe-and-restore-from-scratch for case 1/2, and fail-fast for case 3.

Reply:
@guozhangwang What is your take with regard to "remote stores" that users could implement... I think, we would still need a `clean()` API for this case to make it work.

Reply:
That's a good question: I think rather than adding a `clean()` API, we'd better take our time to resolve the root cause of the issue, to not allow compaction go over specified position (KIP required of course).

Reply:
Why do we need a KIP for this? It's just a topic config we need to change.

Reply:
We do not need the KIP from Streams, but we do need the KIP for Core to add such support that Streams can use.

Reply:
Isn't `min.compaction.lag.ms` providing what we need?

Reply:
KIP-58 is time-diff-based not offset-based, so yes that can be useful but we then have a dependency on wall-clock time drifting. Note that our behavior may be changed to "with the point configurable log compaction, if you still see invalid offset anytime we would just fail directly as it is not expected to happen any more". So if your app may be restarted X days after, you need to set this config to be X days, and hence you may end up with a very large config value which effectively disabled log compaction.

If we can make it time-based (note it is not `time-diff-based`) by specifying "do not compact beyond 2017.12.01 because my app has just caught up to this point" or more concretely be offset-based as "do not compact beyond offset 150" it would be more useful as then Streams app could update this config on-the-fly to still enable log compaction keep moving.

Reply:
Thanks for clarification.

One afterthought: if we assume that people implement a remote store, would we not assume that this remote store is resilient to a failing Kafka Streams instance, and thus, would not need a changelog topic for recovery? For this case, there is no restore and thus, no issue with `InvalidOffsetException`.

However, with regard to exactly-once processing, we would be in a bad state as it's unclear how to roll-back the changes of the in-flight transaction. Only if we would have a changelog topic for transactional store updates, we could wipe out the remote store (given that we add a `clean()` API) and recreate the remote store from the changelog topic (question is still, if we can wipe the remote store only for the data of the specific task -- or if we would need to recreate the remote store by reading all changelog partitions.) Overall, it seems to be a huge undertaking supporting remove stores with exactly-once enabled...